File: 2018_take4/01.lua
local start_time = os.clock()

local lines = {}
for line in io.lines("01_input.txt") do
	table.insert(lines, line)
end

-- part 1
local frequency1 = 0

for i = 1, #lines do
	frequency1 = frequency1 + lines[i]
end

print("Part 1: " .. frequency1)

-- part 2
local counter = 0
local frequency2 = 0
local found_frequencies = {
	["0"] = true
}
local found = false

-- attempt 1 is too slow
--[[
local found_frequencies = { 0 }
local function contains(array, value)
	for i = 1, #array do
		if array[i] == value then
			return true
		end
	end
	return false
end

while (not found) do
	frequency2 = frequency2 + lines[counter % #lines + 1]
	if (contains(found_frequencies, frequency2)) then
		found = true
		break
	end
	--table.insert(found_frequencies, frequency2)
	counter = counter + 1;
end
--]]

-- attempt 2
-- use table as set (with keys as frequencie values and values as true/false/nil)
while (not found) do
	frequency2 = frequency2 + lines[counter % #lines + 1]
	if (found_frequencies[frequency2]) then
		found = true
		break
	end
	found_frequencies[frequency2] = true;
	counter = counter + 1;
end

print("Part 2: " .. frequency2)

local end_time = os.clock()
print("Elapsed time: " .. (end_time - start_time) ..  " seconds")

File: 2018_take7/01.jl
start_time = time()

input_file = "01.txt"

# part 1
frequency = 0

for line in eachline(input_file)
	global frequency += parse(Int, line)
end

println("Day 1 part 1: ", frequency)


# part 2
frequency2 = 0
found_frequencies = Set(frequency2)
found = false

while (!found)
	for line in eachline(input_file)
		global frequency2 += parse(Int, line)

		if (frequency2 in found_frequencies)
			global found = true
			break
		else
			push!(found_frequencies, frequency2)
		end
	end
end

println("Day 1 part 2: ", frequency2)

elapsed_time = time() - start_time
println("Elapsed time: ", elapsed_time, " s")
